Error is thrown when using GET parameters on cloud storage paths

matthiasdv opened the pull request:
## Behaviour

As context, we're storing and reading papermill log data on Azure blob storage. Take the following snippet as an example;

```
import papermill as pm
import scrapbook as sb
import azure.common as ac
import json

papermill_log_output_query_string = "?sig=some-unique-secret-token"

nb_path = "abs://mystorage.blob.core.windows.net/my-actual-notebook.ipynb"

try:
    nb = sb.read_notebook(nb_path + papermill_log_output_query_string)
except ac.AzureMissingResourceHttpError as err:
    logging.info(err)
```

This raises the following error;

```
ValueError: Requires an '.ipynb' file extension. Provided path:...
```

## Expected behaviour

* Azure allows users to provide GET parameters along with the file URL, to enable authorization. In this case an access token. 
* The file should pass, and not raise an extention error


## Suggested solution

* Raise a warning instead of an error
* Or check for GET parameters in the provided urL when cloud storage like Blob or S3 is used. 



I liked the original behaviour of `papermill.record()` which smply raised a warning on the supposedly missing extention. Seeing the warning triggered a developer response of "hey, i've provided a GET parameter in the URL so this is probably fine." 

Not throwing an error possibly hinders developers that provide a faulty path but may no longer notice. Any thoughts on this?

matthiasdv commented:
Update: I've changed the ext check to `contains` instead of `endswith()` since raising a warning still breaks the flow of the program.

MSeal commented:
I think raising a warning instead of an error would be reasonable here as well. I agree that doing contains is better, though we could be a little smarter and do `splitext(urlparse(url).path)[-1].endswith('ipynb')` for the logic to select the warning. If you wanted to change the code to emit a warning here I'll gladly merge it.

MSeal commented:
If you want just this change for now that's fine too. Though a small test to cover the case would be nice either way :)

matthiasdv commented:
I've implemented the suggested url ext check. Added a few simple tests too. One thing i'm unsure about here is that Papermill requires the Azure dependency or will throw a missing dependency exception. 

In case the `Notebook()` call checks the path before actually trying to fetch the object, I believe we can safely ignore the dependency. Which I believe is the case. 

If I'm wrong about this and we actually need to connect to Aure or AWS before checking the path more work would be needed.

matthiasdv commented:
Hmm. Python 2.7 has a module named `urlparse`, and this is renamed to `urllib.parse` in python 3.X. This is what's causing the build to fail.

matthiasdv commented:
fixed this with a conditional import, but unsure if that's the most idiomatic way

MSeal commented on scrapbook/models.py:
```diff
@@ -24,6 +25,12 @@
 from .exceptions import ScrapbookException
 from .utils import kernel_required, deprecated
 
+if sys.version_info > (3, 0, 0):
```
Better to do:

```
try:
    from urllib.parse import urlparse  # Py3
except:
    from urlparse import urlparse  # Py2
```

matthiasdv replied:
Thanks for the info - I've patched the unit test as suggested.

MSeal commented on scrapbook/tests/test_notebooks.py:
```diff
@@ -51,10 +51,24 @@ def test_bad_path():
 
 
+def test_good_ext_for_url():
+    with pytest.raises(PapermillOptionalDependencyException):
```
You shouldn't need this for testing. Do `pip install .[test]` or `pip install .[dev]` (which the test suite does) and you'll install the optional dependencies.

matthiasdv replied:
I'm stuck on getting this test to pass correctly. In my understanding: 

- requirements-dev.txt contains the papermill dependency
- setup.py then specifies `reqs_dev = read_reqs("requirements-dev.txt") + reqs_all` with the latter containing the azure dependency
- tox.ini sets dependencies to .[dev]

However, tox does not install azure dependencies to it's site-packages env folder. Am I missing something?

MSeal replied:
Oh looks like our setuptools isn't correct and missed adding the particular entrypoints for reqs_dev. We need to add `papermill[all]` added to those dependencies. Our tests don't actually ever reach out to remote interfaces so they don't need the specialized io entrypoint to run. But our dev should include them at a minimum and having the test suite install them checks for obvious import issues. `pip install .[azure]` will pull the dependency you're missing. If you want to fix the reqs_dev as well in this PR we can merge that then, or I'll add it later.

matthiasdv replied:
Would you care to elaborate a bit? Something along the lines of: `reqs_dev = read_reqs("requirements-dev.txt").extend(reqs_all)`?

MSeal replied:
probably just change the `papermill` dependencies to `papermill[all]` in requirements-dev.txt. I'll make the change independently, sorry it tripped you up

MSeal commented on scrapbook/tests/test_notebooks.py:
```diff
@@ -51,10 +50,29 @@ def test_bad_path():
+    try:
+        Notebook(url)
+    except TypeError:
+        pass  # this means no valid json is returned by the reader
```
Sorry to nitpick, but instead we should do `mock_read.return_value = <some notebook string>` above and then you won't raise any exception here.

matthiasdv replied:
thanks, that makes more sense. Props for the very nice devops setup too!
